synapse/handlers/auth.py

```diff
@@ -1181,19 +1197,30 @@ async def complete_sso_login(
             respond_with_html(request, 403, self._sso_account_deactivated_template)
             return
 
-        self._complete_sso_login(registered_user_id, request, client_redirect_url)
+        self._complete_sso_login(
+            registered_user_id, request, client_redirect_url, extra_attributes
+        )
 
     def _complete_sso_login(
         self,
         registered_user_id: str,
         request: SynapseRequest,
         client_redirect_url: str,
+        extra_attributes: Optional[JsonDict] = None,
     ):
         """
         The synchronous portion of complete_sso_login.
 
         This exists purely for backwards compatibility of synapse.module_api.ModuleApi.
         """
+        # Store any extra attributes which will be passed in the login response.
+        # Note that this is per-user so it may overwrite a previous value, this
+        # is considered OK since the newest SSO attributes should be most valid.
+        if extra_attributes:
+            self._extra_attributes[registered_user_id] = SsoLoginExtraAttributes(
+                self._clock.time_msec(), extra_attributes,
+            )
```

Comment:
Should the additional information be stored in the database? (I.e. for worker setups?)

Reply:
it'd be really good to avoid having to do that.

In the short term, we can probably document that the /login endpoint needs to hit the same worker as the OIDC callback (ie, the master, I guess?). 

In the longer term, we could use some routing magic (a field in the macaroon?) to make sure that they hit the *same* worker.

Reply:
That's fair. This is actually the case today with SAML anyway (#7530).

Reply:
I added some notes about this in a91fdae606fe522201e57388b12a7384da2b48c0.